File: algorithms/graphics/dda_line/typescript/graphics_algorithms/dda_line.ts
function dda_line(x1: number, y1: number, x2: number, y2: number): Array<Array<number>>{
    let i: number = 0;
    let x: Array<number> = Array();
    let y: Array<number> = Array();

    let x_start: number = Math.round(x1);
    let y_start: number = Math.round(y1);
    let x_end: number = Math.round(x2);
    let y_end: number = Math.round(y2);

    let length: number = Math.max(Math.abs(x_end - x_start), Math.abs(y_end - y_start));

    let dX: number = (x2 - x1) / length;
    let dY: number = (y2 - y1) / length;

    x.push(x1);
    y.push(y1);
    i++;

    while(i < length){
        x.push(x[i - 1] + dX);
        y.push(y[i - 1] + dY);
        i++;
    }

    x.push(x2);
    y.push(y2);

    let result: Array<Array<number>> = Array();

    result.push(x, y);

    return result;
}


export {dda_line};

File: algorithms/graphics/dda_line/python/graphics_algorithms/dda_line.py

def dda_line(x1: float, y1: float, x2: float, y2: float) -> tuple[list[int], list[int]]:
    i = 0
    x = list()
    y = list()
    x_start = round(x1)
    y_start = round(y1)
    x_end = round(x2)
    y_end = round(y2)
    L = max(abs(x_end - x_start), abs(y_end - y_start))
    dX = (x2 - x1) / L
    dY = (y2 - y1) / L
    x.append(float(x1))
    y.append(float(y1))
    i += 1
    while i < L:
        x.append(float(x[i - 1] + dX))
        y.append(float(y[i - 1] + dY))
        i += 1
    x.append(float(x2))
    y.append(float(y2))
    return x, y


if __name__ == "__main__":
    print(dda_line(4, 3, 2, 1))
 
